oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/resourceserver/web/BearerTokenAuthenticationEntryPoint.java

```diff
@@ -0,0 +1,103 @@
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.resourceserver.web;
+
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.security.core.AuthenticationException;
+import org.springframework.security.oauth2.resourceserver.BearerTokenAuthenticationException;
+import org.springframework.security.oauth2.resourceserver.BearerTokenError;
+import org.springframework.security.web.AuthenticationEntryPoint;
+import org.springframework.util.Assert;
+
+/**
+ * An {@link AuthenticationEntryPoint} implementation used to commence authentication of protected resource requests
+ * using {@link BearerTokenAuthenticationFilter}.
+ * <p>
+ * Uses information provided by {@link BearerTokenError} to set HTTP response status code and populate
+ * {@code WWW-Authenticate} HTTP header.
+ *
+ * @author Vedran Pavic
+ * @since 5.1
+ * @see BearerTokenAuthenticationException
+ * @see BearerTokenError
+ * @see <a href="https://tools.ietf.org/html/rfc6750#section-3" target="_blank">RFC 6750 Section 3: The WWW-Authenticate
+ * Response Header Field</a>
+ */
+public class BearerTokenAuthenticationEntryPoint implements AuthenticationEntryPoint {
+
+	private String realmName;
+
+	@Override
+	public void commence(HttpServletRequest request, HttpServletResponse response,
+			AuthenticationException authException) throws IOException {
+		HttpStatus httpStatus;
+		Map<String, String> authParamAttributes = new LinkedHashMap<>();
+		if (this.realmName != null) {
+			authParamAttributes.put("realm", this.realmName);
+		}
+		if (authException instanceof BearerTokenAuthenticationException) {
+			BearerTokenError error = ((BearerTokenAuthenticationException) authException).getError();
+			httpStatus = error.getHttpStatus();
+			authParamAttributes.put("error", error.getErrorCode());
+			String description = error.getDescription();
+			if (description != null) {
+				authParamAttributes.put("error_description", description);
+			}
+			String uri = error.getUri();
+			if (uri != null) {
+				authParamAttributes.put("error_uri", uri);
+			}
+			String scope = error.getScope();
+			if (scope != null) {
+				authParamAttributes.put("scope", scope);
+			}
+		}
+		else {
+			httpStatus = HttpStatus.UNAUTHORIZED;
+		}
+		response.addHeader("WWW-Authenticate", buildWwwAuthenticate(authParamAttributes));
+		response.sendError(httpStatus.value(), httpStatus.getReasonPhrase());
+	}
+
+	/**
+	 * Set the realm name.
+	 * @param realmName the realm name
+	 */
+	public void setRealmName(String realmName) {
+		Assert.hasText(realmName, "realmName must not be null");
+		this.realmName = realmName;
+	}
+
+	private static String buildWwwAuthenticate(Map<String, String> authParamAttributes) {
```

Comment:
It's a utility method, any reason for it not to be static?

Reply:
Less typing. :)

Really, though, I haven't seen someone take the time to add a static keyword to a private method. Do you see value in constraining its access to the instance?

Reply:
As this is considered a utility method, I tend to make those static the same way it would be in a public utility class. It's a signal that method shouldn't mess with the state.

Regarding constraining the access, if you expose it then it's out there so you can't change it easily if you need. Until there's a good reason to expose it as public, we'd better keep it internal.

Reply:
I agree we should keep it `private` (never expose until we have/need to rule). However, I'm not convinced that it should be static. Utility methods are typically in utility classes.

Reply:
@vpavic, I'm actually fine either way (static or no), but allow me to clarify that I wasn't advocating changing it to public. By "access", I meant what you said--that the method can't access instance state due to it being static.

Reply:
To clarify: I've introduced `buildWwwAuthenticate` method to avoid cluttering `commence` method with `WWW-Authenticate` header value building logic. As a result in `commence` we only have:

```
response.addHeader("WWW-Authenticate", buildWwwAuthenticate(authParamAttributes));
```

So, this makes `buildWwwAuthenticate` a utility/helper method by its purpose as it doesn't use or modify instance state. As such, making it static clearly communicates its nature at both place of invocation (as IDE will show it in italics, which improves code readability & understanding) and to whoever touches this piece of code next time.

This pattern is quite common, and quick search (`grep -r "private static \w\{1,\} \w\{1,\}(" . | wc -l`) reveals number of occurrences well in triple digits in Spring Framework/Boot/Security codebases.

Reply:
> So, this makes buildWwwAuthenticate a utility/helper method by its purpose as it doesn't use or modify instance state. As such, making it static clearly communicates its nature

There are far more `private` (or `protected`) utility/helper instance methods that do not modify state so the same argument can be made...regarding `private static` utility/helpers. My preference is to remove the `static` modifier.

>  introduced buildWwwAuthenticate method to avoid cluttering commence method with WWW-Authenticate header value building logic

I would also be ok with removing this method altogether as I don't feel it would clutter the `commence` method.